example/merge.cpp

```diff
@@ -0,0 +1,129 @@
+//
+// Copyright (c) 2019 Vinnie Falco (vinnie.falco@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/json
+//
+
+#include <boost/json/object.hpp>
+#include <boost/json/serialize.hpp>
+#include <cstdlib>
+#include <iostream>
+
+namespace boost {
+namespace json {
+
+/** Merge the elements from one object into another
+
+    For each key in `src` this function attempts
+    to move the entire key/value pair into `dest`
+    if the key is not already present. Each
+    element successfully moved is erased from
+    `src`, making this operation destructive
+    to `src`.
+
+    @par Complexity
+    Linear in `src.size()` best case,
+    quadratic in `dest.size()+src.size()` worst case.
+
+    @par Exception Safety
+    Basic guarantee.
+    Calls to allocate may throw.
+
+    @param dest The object to merge into
+
+    @param src The object to remove elements from
+*/
+object&
+merge_move(
+    object& dest,
+    object&& src)
+{
+    dest.reserve(
+        dest.size() + src.size());
+    auto it = src.end();
+    auto const begin = src.begin();
+    while(it != begin)
+    {
+        --it;
+        auto result =
+            dest.insert(pilfer(*it));
+        if(result.second)
+            src.erase(it);
+    }
+    return dest;
+}
```

Comment:
This should be almost equivalent to your code:
```c++
object&
merge_objects(
    object& dest,
    object src)
{
    dest.insert(
        std::make_move_iterator(src.begin()),
        std::make_move_iterator(src.end()));
    return dest;
}
```
And it's basically one library function (with a little help of the standard library). `merge_copy` is the same function, only the caller capies rather than moves `src`.